storage/ctab/src/main/java/org/openscience/cdk/io/RdfileReader.java

```diff
@@ -0,0 +1,533 @@
+package org.openscience.cdk.io;
+
+import org.openscience.cdk.exception.CDKException;
+import org.openscience.cdk.interfaces.IChemObject;
+import org.openscience.cdk.interfaces.IChemObjectBuilder;
+import org.openscience.cdk.tools.ILoggingTool;
+import org.openscience.cdk.tools.LoggingToolFactory;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Iterating reader for RDFiles.
+ *
+ * @see RdfileRecord
+ * @author Uli Fechner
+ */
+public class RdfileReader implements Closeable, Iterator<RdfileRecord> {
+    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(RdfileReader.class);
+    static final String RDFILE_VERSION_1 = "$RDFILE 1";
+    static final String DATM = "$DATM";
+    static final String REACTION_FMT = "$RFMT";
+    static final String REACTION_INT_REG = "$RIREG";
+    static final String REACTION_EXT_REG = "$REREG";
+    static final String RXNFILE_START = "$RXN";
+    static final String MOLECULE_FMT = "$MFMT";
+    static final String MOLECULE_INT_REG = "$MIREG";
+    static final String MOLECULE_EXT_REG = "$MEREG";
+    static final String MOLFILE_START = "$MOL";
+    static final String M_END = "M  END";
+    static final String DTYPE = "$DTYPE";
+    static final String DATUM = "$DATUM";
+    static final String LINE_SEPARATOR_NEWLINE = "\n";
+    static final String PLUS_SIGN = "+";
+    private static final Pattern DTYPE_KEY = Pattern.compile("\\A\\" + DTYPE + " (.+)\\Z");
+    private static final Pattern DATUM_VALUE_FIRSTLINE = Pattern.compile("\\A\\" + DATUM + " (.{73}\\+|.{1,72})\\Z");
+    private static final Pattern MDL_CTAB_VERSION = Pattern.compile("(V[23]0{3})\\Z");
+    private static final Pattern MDL_RXN_VERSION = Pattern.compile("\\A\\$RXN ?(V30{3})?\\Z");
+
+    private boolean headerRead;
+    private String previousLine;
+    private int lineCounter;
+    private RdfileRecord nextRecord;
+    private boolean endOfFile;
+    private final boolean skipRecordsOnError;
+    private final BufferedReader bufferedReader;
+    private final IChemObjectBuilder chemObjectBuilder;
+
+    public RdfileReader(InputStream in, IChemObjectBuilder chemObjectBuilder) {
+        this(in, chemObjectBuilder, true);
+    }
+
+    public RdfileReader(InputStream in, IChemObjectBuilder chemObjectBuilder, boolean skipRecordsOnError) {
+        this(new InputStreamReader(in, StandardCharsets.UTF_8), chemObjectBuilder, skipRecordsOnError);
+    }
+
+    public RdfileReader(Reader reader, IChemObjectBuilder chemObjectBuilder) {
+        this(reader, chemObjectBuilder, true);
+    }
+
+    public RdfileReader(Reader reader, IChemObjectBuilder chemObjectBuilder, boolean skipRecordsOnError) {
+        this.bufferedReader = new BufferedReader(reader);
```

Comment:
Same as above, you can check if it is a BufferedReader first.

Reply:
Will do.